articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+The structure contains information about the `Subscription`, such as its QoS, its topic name and the type of its callback, and a weak pointer for the `Subscription` object.
+An `uint64_t sub_id` unique within the `rclcpp::Context` is assigned to the `Subscription`.
+The `IntraProcessManager` contains a `std::map<uint64_t, SubscriptionInfo>` object where it is possible to retrieve the `SubscriptionInfo` of a specific `Subscription` given its id.
+There is also an additional structure `std::map<uint64_t, std::pair<std::set<uint64_t>, std::set<uint64_t>>>`.
+The key of the map is the unique id of a `Publisher` and the value is a pair of sets of ids.
+These sets contain the ids of the `Subscription`s that can communicate with the `Publisher`.
+We have two different sets because we want to differentiate the `Subscription`s depending on whether they request ownership of the received messages or not (note that this decision is done looking at their buffer, since the `Publisher` does not have to interact with the `Subscription` callback).
+6. The `SubscriptionIntraProcessWaitable` object is added to the list of Waitable interfaces of the node through `node_interfaces::NodeWaitablesInterface::add_waitable(...)`.
+It is added to the same callback group used for the standard inter-process communication of that topic.
+
+### Publishing only intra-process
+
+#### Publishing unique_ptr
+
+1. User calls `Publisher::publish(std::unique_ptr<MessageT> msg)`.
+2. `Publisher::publish(std::unique_ptr<MessageT> msg)` calls `IntraProcessManager::do_intra_process_publish(uint64_t pub_id, std::unique_ptr<MessageT> msg)`.
+3. `IntraProcessManager::do_intra_process_publish(...)` uses the `uint64_t pub_id` to call `IntraProcessManager::get_subscription_ids_for_pub(uint64_t pub_id)`.
+This returns the ids correspondin to `Subscription`s that have a QoS compatible for receiving the message.
+These ids are divided into two sublists, according to the data-type that is stored in the buffer of each `Susbscription`: requesting ownership (`unique_ptr<MessageT>`) or accepting shared (`shared_ptr<MessageT>`, but also `MessageT` since it will copy data in any case).
+4. The message is "added" to the ring buffer of all the items in the lists.
+The `rcl_guard_condition_t` member of `SubscriptionIntraProcessWaitable` of each `Subscription` is triggered (this wakes up `rclcpp::spin`).
+
+The way in which the `std::unique_ptr<MessageT>` message is "added" to a buffer, depends on the type of the buffer.
+
+ - `BufferT = unique_ptr<MessageT>` The buffer receives a copy of `MessageT` and has ownership on it; for the last buffer, a copy is not necessary as ownership can be transferred.
+ - `BufferT = shared_ptr<const MessageT>` Every buffer receives a shared pointer of the same `MessageT`; no copies are required.
+ - `BufferT = MessageT` A copy of the message is added to every buffer.
```

Comment:
The original idea behind that was to use it for very small data (i.e. base types) especially in case of high number of subscriptions. In this way you avoid the overhead of synchronizing the shared pointers and of having many threads accessing the same memory address.

I did some quick tests, but I haven't seen noticeable improvements. 
There are obviously alternatives, such as using unique pointers that are already supported.